python/src/opendp/mod.py

```diff
@@ -210,3 +218,84 @@ def make_chain_tt_multi(self, *transformations):
     def to_str(self, data):
         string = self.data.to_string(data)
         return c_char_p_to_str(string)
+
+    def get_first(self, list):
+        return list[0] if list else 0
+
+    def get_ffi_type_name(self, val):
+        if isinstance(val, int):
+            return "i32"
+        elif isinstance(val, float):
+            return "f64"
+        elif isinstance(val, str):
+            return "String"
+        elif isinstance(val, list):
+            element_type_name = self.get_ffi_type_name(self.get_first(val))
+            return f"Vec<{element_type_name}>"
+        else:
+            raise Exception("Unknown type", type(val))
+
+    def to_raw(self, val):
+        if isinstance(val, int):
+            ptr, len_ = ctypes.byref(ctypes.c_int32(val)), 1
+        elif isinstance(val, float):
+            ptr, len_ = ctypes.byref(ctypes.c_double(val)), 1
+        elif isinstance(val, str):
+            ptr, len_ = ctypes.c_char_p(val.encode()), len(val) + 1
+        elif isinstance(val, list):
+            first = self.get_first(val)
+            if isinstance(first, int):
+                element_type = ctypes.c_int32
+            elif isinstance(first, float):
+                element_type = ctypes.c_double
+            else:
+                raise Exception("Unknown element type", type(first))
+            array = (element_type * len(val))(*val)
+            ptr, len_ = array, len(val)
+        else:
+            raise Exception("Unknown type", type(val))
+        return ctypes.byref(FfiSlice(ctypes.cast(ptr, ctypes.c_void_p), len_))
+
+    def py_to_obj(self, val):
+        type_name = self.get_ffi_type_name(val)
+        type_args = f"<{type_name}>".encode()
+        raw = self.to_raw(val)
+        return self.data.object_new(type_args, raw)
+
+    def from_raw(self, type_name, raw):
+        if type_name == "i32":
+            return ctypes.cast(raw.contents.ptr, ctypes.POINTER(ctypes.c_int32)).contents.value
+        elif type_name == "f64":
+            return ctypes.cast(raw.contents.ptr, ctypes.POINTER(ctypes.c_double)).contents.value
+        elif type_name == "String":
+            return ctypes.cast(raw.contents.ptr, ctypes.c_char_p).value.decode()
+        elif type_name.startswith("Vec<"):
+            if type_name == "Vec<i32>":
+                array = ctypes.cast(raw.contents.ptr, ctypes.POINTER(ctypes.c_int32))
+            elif type_name == "alloc::vec::Vec<f64>":
```

Comment:
Did you mean to leave `alloc::vec::` on here? This branch is unreachable.

Reply:
Good catch! I'll fix that, and add a test for that case.